docs/configuration.md
# Configuration

## File structure

All Marlinraker-related files will be stored in a folder called ``marlinraker_files``.
It will be located next to the directory Marlinraker is running in. For example, on the
pre-built image Marlinraker is running in ``/pi/home/marlinraker``, so it will be located
at ``/pi/home/marlinraker_files``. It contains the following subdirectories and files:

``config/``<br>
This is where config files are stored. This folder will be exposed by the API and contents will
be editable in your web interface of choice. See [Configuration reference](#configuration-reference) for more info.

``gcodes/``<br>
User-uploaded G-code files will be stored in this folder. It will too be exposed by the API.

``logs/``<br>
Logs are stored here. The most recent log file will always be called ``marlinraker.log``.
Log files will be rotated after they reach a size of 1 megabyte. If there are more than
5 files, old logs will be deleted.

``update_scripts/``<br>
Update scripts are located here. See [Update scripts](advanced/update-manager.md) for more info.

``www/``<br>
This is where the web interface will be located. The contents of this folder will be
statically served.

## Configuration reference

Marlinraker uses [TOML](https://toml.io/) as its format for configuration files. Marlinraker 
will always load the `marlinraker.toml` config file. From there on it is possible to
include other configuration files using the `include` directive:

```toml
#<include file/to/include.toml>
```

This can be used to efficiently organize configuration files and swap them out if needed.
That's why by default all printer-related settings are stored in `printer.toml` and
included in the main config file.

[Here](https://github.com/pauhull/marlinraker/tree/master/config/printers) you can find 
pre-made printer configuration files. If you find a configuration matching your printer
you can copy its contents into `printer.toml`.

### General settings

```toml
[web]
port = 7125
# The port Marlinraker will run on. Note that this is
# not necessarily the port your web interface will run on.
cors_domains = []
# Domains that have acess to the Marlinraker API. If there are
# no domains specified, CORS is disabled. Ideally you should
# enter all the domains you want to access Marlinraker from, e.g.:
# ["http://my-printer.local", "http://1.2.3.4"]

[serial]
port = "auto"
# Serial port to connect to. Use "auto" for auto-detecting
# any connected printer. Can alternatively be a USB port
# like "/dev/ttyUSB0" on Linux or "COM1" on Windows.
baud_rate = "auto"
# Baud rate to use for the USB connection. Use "auto" for
# auto-detection. Can alternatively be a integer value like
# 115200 or 250000.
max_connection_attempts = 5
# Number of connection attempts before connection is terminated.
# Can be useful for printers that reset after a connection
# has been opened and need some time to start up.
connection_timeout = 5000
# Time in ms after a connection times out if no response was
# detected.

[misc]
octoprint_compat = true
# Set to true to enable OctoPrint API emulation. This is useful
# if you want to upload G-code to your printer directly from
# your slicer.
extended_logs = false
# Output debug information to logs. Only recommended for developers
# and advanced users, can negatively impact performance.
report_velocity = false
# Poll position in very short intervals to approximately compute
# current toolhead velocity. This can cause stuttering on older
# versions of Marlin or slow MCUs.
allowed_services = ["marlinraker", "crowsnest", "MoonCord",
    "moonraker-telegram-bot", "KlipperScreen", "sonar", "webcamd"]
# List of services that can be (re-)started or stopped from the API.
```

### Printer-specific settings
```toml
[printer]
bed_mesh = false
# Set to true if your printer has bed mesh leveling.
print_volume = [220, 220, 240]
# x, y and z print volume of your printer.

[printer.extruder]
min_temp = 0
# Extruder minimum temperature in C.
max_temp = 250
# Extruder maximum temperature in C. Note that this setting
# will not actually prevent your printer from exceeding
# this temperature!
min_extrude_temp = 180
# Minimum extrude temperature in C. As with min_temp and max_temp,
# this will not actually limit your machine.
filament_diameter = 1.75
# Filament diameter in mm (For display purposes only).

[printer.heater_bed]
min_temp = 0
# See printer.extruder.min_temp
max_temp = 100
# See printer.extruder.max_temp

[printer.gcode]
send_m73 = true
# Whether to send M73 (Print progress) G-codes to your printer while
# printing. Note that some slicers already include M73 G-codes in sliced
# files. If you notice that the progress indicator is jumping between values,
# set this option to false.
```

### Macros

Marlinraker allows you to define macros. Macros allow you to extend your
printer's functionality with own G-code commands. A macro might look like
this:

```toml
[macros.example]
rename_existing = "example_base"
# Optional. If this macro overrides an already existing macro, this option
# can be used to rename it.
gcode = """
G28 ; Home printer
M118 E1 Hi mom!
"""
# The actual G-code that will be run when executing this macro. 
```

You likely want to redefine certain macros to better fit your printer. Macros
you can redefine include:

- `cancel_print`
- `pause`
- `resume`
- `sdcard_reset_file`
- `start_print`

Advanced users might want to have a look at [Macro scripting](advanced/macro-scripting.md).

CHANGELOG.md
# Changelog

All notable changes to Marlinraker will be documented in this file.

## 0.2.3-alpha - 2023-04-10

### Bug Fixes

- Add misc.allowed_services to config/marlinraker.toml and docs ([b3d4abd](b3d4abdcbaf019286a0ca141db06729c7b3befbb))
- Don't crash when udevadm isn't found ([223d6e9](223d6e97eee9fce9779955f2381ce6c4660ff3b2))
- Fix Docker image ([9db7b09](9db7b0948c82812d0eeba57569f5039444f401fd))
- Fix "Printer initialization took too long" on Prusa printers ([b76f0c2](b76f0c2241e181cfb6712ac1d5af858ca8e02e54))
- Show velocity while printing when "report_velocity" is enabled ([70a4866](70a4866f3404ae4c37fe98474dc8226d9f17fb6e))

### Features

- Add and document 'cors_domains' configuration option ([b2965e3](b2965e3359a4485c071d49b3a6f7a61f2977f025))
- Add printer.extruder.filament_diameter property ([cda5adb](cda5adbd689228fbcb23651bcd85e50f4c6c342c))



src/config/Config.ts
import fs from "fs-extra";
import path from "path";
import { logger } from "../Server";
import TOML from "@iarna/toml";
import Utils, { TVec3 } from "../util/Utils";

interface IFileInfo {
    filename: string;
    sections: string[];
}

class Config {

    public klipperPseudoConfig: object;
    public readonly config: unknown;
    public readonly files: IFileInfo[];
    public readonly warnings: string[];
    private readonly cache: Record<string, unknown | undefined>;
    private readonly configFile: string;
    private readonly isConfigMalformed: boolean;

    public constructor(configFile: string) {
        this.warnings = [];
        this.cache = {};
        this.configFile = path.resolve(configFile);
        this.isConfigMalformed = false;
        [this.config, this.files] = this.load();
        this.klipperPseudoConfig = this.generateKlipperConfig();
        logger.info("Config loaded");
        logger.debug(`Config:\n${JSON.stringify(this.config, null, 2)}`);
        logger.debug(`Klipper pseudo config:\n${JSON.stringify(this.klipperPseudoConfig, null, 2)}`);
    }

    private resolve(filepath: string, resolvedSoFar: string[] = []): [object, IFileInfo[]] {

        if (!fs.pathExistsSync(filepath)) {
            throw new Error(`${filepath} does not exist`);
        }

        const originalContent = fs.readFileSync(filepath).toString("utf-8");
        const resolvedContent = TOML.parse(originalContent.replace("\\", "\\\\"));
        resolvedSoFar.push(filepath);

        const fileInfo = [{
            filename: path.relative(path.dirname(this.configFile), filepath),
            sections: Utils.getDeepKeys(resolvedContent)
        }];

        const regex = /^#<include +([\w\-. /\\]+?)>.*$/gmi;
        let match;
        while ((match = regex.exec(originalContent)) !== null) {
            const file = match[1];
            const target = path.resolve(path.dirname(filepath), file);

            if (resolvedSoFar.some((resolved) => !path.relative(resolved, target))) {
                throw new Error(`Cyclic reference to ${path.basename(target)}`);
            }

            try {
                const [resolvedImport, importFileInfo] = this.resolve(target, resolvedSoFar);
                Object.assign(resolvedContent, resolvedImport);
                fileInfo.push(...importFileInfo);
            } catch (e) {
                const lineNum = originalContent.substring(0, match.index).split(/\r?\n/).length;
                logger.error(`Cannot resolve import ${path.basename(target)} in ${path.basename(filepath)}, line ${lineNum}: ${Utils.errorToString(e)}`);
            }
        }

        return [resolvedContent, fileInfo];
    }

    private load(): [unknown, IFileInfo[]] {
        try {
            return this.resolve(this.configFile);
        } catch (e) {
            logger.error("Loading config failed:");
            logger.error(e);
        }
        return [{}, []];
    }

    public getObject(property: string, defaultValue: object): object {
        return this.getGeneric<object>(property, defaultValue, (x): x is object => typeof x === "object");
    }

    public getString(property: string, defaultValue: string): string {
        return this.getGeneric<string>(property, defaultValue, (x): x is string => typeof x === "string");
    }

    public getStringIfExists(property: string, defaultValue: string | null): string | null {
        return this.getGeneric<string | null>(property, defaultValue, (x): x is string | null => typeof x === "undefined" || typeof x === "string") ?? null;
    }

    public getNumber(property: string, defaultValue: number): number {
        return this.getGeneric<number>(property, defaultValue, (x): x is number => typeof x === "number");
    }

    public getNumberIfExists(property: string, defaultValue: number | null): number | null {
        return this.getGeneric<number | null>(property, defaultValue, (x): x is number | null => typeof x === "undefined" || typeof x === "number") ?? null;
    }

    public getStringOrNumber(property: string, defaultValue: string | number): string | number {
        return this.getGeneric(property, defaultValue, (x): x is string | number => typeof x === "string" || typeof x === "number");
    }

    public getBoolean(property: string, defaultValue: boolean): boolean {
        return this.getGeneric<boolean>(property, defaultValue, (x): x is boolean => typeof x === "boolean");
    }

    public getStringArray(property: string, defaultValue: string[]): string[] {
        return this.getGeneric<string[]>(property, defaultValue, (x): x is string[] => Array.isArray(x) && x.every((el) => typeof el === "string"));
    }

    public getGeneric<T>(property: string, defaultValue: T, check: (x: unknown) => x is T): T {
        if (this.cache[property] !== undefined) {
            return this.cache[property] as T;
        }
        let value = this.config;
        for (const next of property.split(".")) {
            value = (value as Record<string, unknown> | undefined)?.[next];
        }
        if (value === undefined && defaultValue !== null) {
            logger.warn(`Warning: "${property}" was not found in ${this.configFile}`);
            logger.warn(`Using default value ${JSON.stringify(defaultValue)} for property ${property}`);
            this.warnings.push(`Property "${property}" is missing. Using default value ${JSON.stringify(defaultValue)}.`);
            value = defaultValue;
        } else if (!check(value)) {
            logger.warn(`Warning: "${property}" is of wrong type (${typeof value})`);
            if (defaultValue !== null) {
                logger.warn(`Using default value ${JSON.stringify(defaultValue)} for property ${property}`);
                this.warnings.push(`Property "${property}" is of wrong type (${typeof value}). Using default value ${JSON.stringify(defaultValue)} instead.`);
            } else {
                this.warnings.push(`Property "${property}" is of wrong type (${typeof value}).`);
            }
            value = defaultValue;
        }
        this.cache[property] = value;
        return value as T;
    }

    private generateKlipperConfig(): object {
        const printVolume = this.getGeneric<TVec3>("printer.print_volume",
            [220, 220, 240], (x): x is TVec3 =>
                typeof x === "object" && Array.isArray(x) && x.length === 3
        );
        const klipperConfig = {
            extruder: {
                min_temp: this.getNumber("printer.extruder.min_temp", 0),
                max_temp: this.getNumber("printer.extruder.max_temp", 250),
                min_extrude_temp: this.getNumber("printer.extruder.min_extrude_temp", 180),
                filament_diameter: this.getNumber("printer.extruder.filament_diameter", 1.75)
            },
            heater_bed: {
                min_temp: this.getNumber("printer.heater_bed.min_temp", 0),
                max_temp: this.getNumber("printer.heater_bed.max_temp", 60)
            },
            bed_mesh: {
                mesh_min: "0, 0",
                mesh_max: `${printVolume[0]}, ${printVolume[1]}`
            },
            virtual_sdcard: {},
            pause_resume: {},
            display_status: {}
        };
        const override = (this.config as Record<string, unknown>).klipper_pseudo_config;
        if (override === Object(override)) { // check if is object
            Object.assign(klipperConfig, override);
        }
        return klipperConfig;
    }
}

export { IFileInfo };
export default Config;